### DISCUSSION TITLE
Change stabilizer task time

### DISCUSSION THREAD

**Discussion by xkhainguyen** [bitcraze/discussions#793] [Q&A] [answered]
Hi,

Is it possible to make the stabilizer task run at 500Hz instead of 1000Hz?
It seems like RATE_MAIN_LOOP won't affect the stabilizer task, it serves as the base for other functions. I feel like the `sensorsWaitDataReady()` is the one that decides the frequency of it.

Last time, we tried to put my controller into a new task but that did slow things down because other higher-frequency tasks (like stabilizer) will just keep interrupting it. We think that keeping the controller inside the stabilizer task is better as it focuses all resources at once. However, we need to change the stabilizer task to our controller frequency to get the best effect. 

Additionally, will my controller function block the tick (timing) of RTOS as well? Sometimes even though my controller computation often takes more than 1 ms, it does not warn me about any rate drops.

Could you help me to resolve these questions?
Thanks.

**Comment by krichardsson** [comment]
Hi!

The stabilizer task is running at 1000Hz and is controlled (like you say) by the `sensorsWaitDataReady()` function, that in turn is controlled by the IMU.
The define `RATE_MAIN_LOOP` is mainly used to calculate other rates, not control the stabilizer loop rate.

> Last time, we tried to put my controller into a new task but that did slow things down because other higher-frequency tasks (like stabilizer) will just keep interrupting it. We think that keeping the controller inside the stabilizer task is better as it focuses all resources at once. However, we need to change the stabilizer task to our controller frequency to get the best effect.

In theory you could change the frequency of the stabilizer loop, but I think there might be implications in other parts of the system that are hard to predict (like the estimator). You can always try :-)

If you want to run the controller in the stabilizer loop, you will have to make sure it does not take too long. One option is to split the control work up to do one part of the work in every iteration of the stabilizer loop and get the final result every 20 iterations (or similar). The key is that the resources in the system is limited. I still think a task is the easier solution (as you don't have to manage the chopping up of the problem), this should not add a lot of overhead, the basic problem is the lack of resources. 

> Additionally, will my controller function block the tick (timing) of RTOS as well? Sometimes even though my controller computation often takes more than 1 ms, it does not warn me about any rate drops.

The ticks in FreeRTOS are handled at a very low level and they should continue, even if your code takes longer to run. There is no warning though about rates in FreeRTOS, you have to implement that your self, see the [check in the stabilizer loop](https://github.com/bitcraze/crazyflie-firmware/blob/41a9c97569e565e927a0fe3b8f032dbcdb03aa1b/src/modules/src/stabilizer.c#L330) for inspiration.

Also note that tasks behave differently from threads in linux (for instance). You might want to read up on tasks in the FreeRTOS documentation to understand how they are interrupted, priorities and so on.

**Reply by xkhainguyen** [reply]
Thank you for your response!

We did move the controller to the task (priority 5) but it takes 2x more computation time than the previous approach because as I said, the stabilizer task keeps interrupting the controller task. 

So, instead of restricting the controller computation to the stabilizer task at 1000Hz, we want to lower the stabilizer task to 500Hz so we get 2ms for everything.

> In theory you could change the frequency of the stabilizer loop, but I think there might be implications in other parts of the system that are hard to predict (like the estimator). You can always try :-)

I would love to know the correct way to do that. Does the `sensorsWaitDataReady()` have something with the sensor driver? We use the basic CF2.1 with OptiTrack.
We now try to count the `sensorsWaitDataReady()` runs to set the frequency. Is it even valid?

> The ticks in FreeRTOS are handled at a very low level and they should continue, even if your code takes longer to run. There is no warning though about rates in FreeRTOS, you have to implement that your self, see the [check in the stabilizer loop](https://github.com/bitcraze/crazyflie-firmware/blob/41a9c97569e565e927a0fe3b8f032dbcdb03aa1b/src/modules/src/stabilizer.c#L330) for inspiration.

I mean I did not get that many Kalman filter and stabilizer rate drop warnings though my controller computation time often exceeds 1ms.

**Comment by krichardsson** [comment, marked as answer]
> We did move the controller to the task (priority 5) but it takes 2x more computation time than the previous approach because as I said, the stabilizer task keeps interrupting the controller task.

Some random thoughts related to tasks and priorities that perhaps will help you forward:

Priority 5 sounds a bit too high, I would have tried the same level as the kalman estimator, priority 2 or perhaps 3.
In general terms, priority 5 is for the stabilizer loop, priority 4 for sensors and priority 3 for lower priority sensor crunching.

It is important to understand that the CPU is a limited resource that must be shared by all the tasks that runs in the system, changing the priority of tasks does not create more CPU power, it just changes which task that gets CPU time. The problem of running too much work at priority 5 is that you might starve sensors and other important processes which can cause problems. 

You mention that you used priority 5 for your controller task, this is the same priority as the stabilizer loop. The OS will switch between tasks at the same priority level, which means that running your controller in a task at level 5 is very similar to running it directly in the stabilizer loop.

> So, instead of restricting the controller computation to the stabilizer task at 1000Hz, we want to lower the stabilizer task to 500Hz so we get 2ms for everything.

You can use the "Task dump" button in the python client/console tab to get a list of how much CPU the tasks in the system uses. Note that the numbers are the average since the previous measurement.

As you are low on CPU power you might want to examine where it is possible to reduce the load. You could for instance check how often you push mocap data to the Crazyflie. Reducing the rate will offload both communication and estimator tasks. The Crazyflie will fly with mocap update rates as low as 5Hz, but you might want more for your experiments? Anyway this can be worth looking into.
You might also consider the required update rate of your controller, what is actually required? The standard PID controller in the Crazyflie runs [the attitude part at 500Hz](https://github.com/bitcraze/crazyflie-firmware/blob/41a9c97569e565e927a0fe3b8f032dbcdb03aa1b/src/modules/src/controller/controller_pid.c#L63) and the [position part at 100Hz](https://github.com/bitcraze/crazyflie-firmware/blob/41a9c97569e565e927a0fe3b8f032dbcdb03aa1b/src/modules/src/controller/controller_pid.c#L93).

Running the stabilizer task at 500Hz will save some CPU power, but unfortunately I don't think it will make a big difference. The Stabilizer task uses around 5% CPU so reducing it to half would give you around 2% CPU more. 

If you want to run your controller every second time you can use the `RATE_DO_EXECUTE` macro like we do in the [pid controller](https://github.com/bitcraze/crazyflie-firmware/blob/41a9c97569e565e927a0fe3b8f032dbcdb03aa1b/src/modules/src/controller/controller_pid.c#L63)

> I would love to know the correct way to do that. Does the sensorsWaitDataReady() have something with the sensor driver? We use the basic CF2.1 with OptiTrack.
We now try to count the sensorsWaitDataReady() runs to set the frequency. Is it even valid?

Yes, `sensorsWaitDataReady()` is essentially a semaphore that is release when new gyro data is available (I think). The Mocap data (optitrack) is pushed into the estimator and is not handled in the stabilizer loop. 

I think it would work to simply add a second call to `sensorsWaitDataReady()` at the top of the loop:
```C
  while(1) {
    // The sensor should unlock at 1kHz
    sensorsWaitDataReady();
    // Wait for the next release to run the stabilizer loop at 500Hz
    sensorsWaitDataReady();

    // You might have to increase `stabilizerStep` one extra time as well. I'm not sure how this will effect other parts of the system though
    stabilizerStep++;

    // update sensorData struct (for logging variables)
    sensorsAcquire(&sensorData);
   // ... the rest of the stabilizer loop code
```
